### PR TITLE
Add Sensor Device class and allow device factory to make instance

### PR DESCRIPTION
**Pull request by mklewandowski**
Add a new D2PIOSensorDevice class which can be instantiated from the JavaDeviceFactory.  Add new sensor cases to verniersensor.java.  Build new list of Go Direct sensor ids in SensorID.java.  Add a d2pio dependency to the pom.xml file.  Ensure that the sensor-connector app can properly identify a Go Direct device via its USB product id.

Notes:
- some of these changes are included in previous PRs for ease of review, but are required here as well. 
- additional changes are required to work in sensor-connector and are located here: https://github.com/concord-consortium/sensor-connector/tree/161195811-go-direct-sensor-connector-integration

### CODE REVIEW

**Comment by mklewandowski** [comment]
@kswenson I think I missed adding you as a reviewer when I created the PR.  This is the last branch that we were looking at and tweaking before we left today.

File: sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+	public String getVendorName() {
+		return "Vernier";
+	}
+
+	public String getDeviceName() {
+		if(currentGoDirectDevice != null){
+			return currentGoDirectDevice.getDeviceLabel();
+		}
+
+		return "GoDirect";
```
Comment (kswenson) [inline sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java]: What's the relationship between the device name and the device label?

File: sensor-vernier/src/main/java/org/concord/sensor/vernier/VernierSensor.java [hunk 2]
```diff
@@ -363,34 +435,34 @@ public int setupSensor(int sensorId, SensorRequest request)
...
-				// we keep this artificially low so we don't restrict 
+				setType(QUANTITY_TEMPERATURE);
+
+				// we keep this artificially low so we don't restrict
 				// malformed requests which claim to require small step sizes
-				setStepSize(0.01f); 
+				setStepSize(0.01f);
 				setCalibration(temperatureCalibration);
 				break;
 			case SensorID.THEROCOUPLE:
```
[the remark this answers is not in the document]
Comment (mklewandowski) [reply sensor-vernier/src/main/java/org/concord/sensor/vernier/VernierSensor.java]: I assume that's meant to be THERMOCOUPLE....

File: sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+	}
+
+	@Override
+	protected boolean hasNonAutoIdSensors() {
+		return false;
+	}
+
+	public ExperimentConfig configure(ExperimentRequest request) {
+		// FIXME this should reject raw* requests if that don't work with the attached device
+		//   for example raw_*2 shouldn't work with the goTemp or goMotion
```
[the remark this answers is not in the document]
Comment (mklewandowski) [reply sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java]: I'll remove.  I don't think it is relevant.

File: sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+import org.concord.sensor.vernier.VernierSensorDevice;
+import javax.swing.JOptionPane;
+
+public class D2PIOSensorDevice extends AbstractSensorDevice implements
+		VernierSensorDevice
+{
+	D2PIOLibrary d2pio;
+	String errorMessage;
+	D2PIOSensor currentGoDirectDevice;
+  private static final int MAX_D2PIO_CHANNELS = 32;
```
[the remark this answers is not in the document]
Comment (mklewandowski) [reply sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java]: Looks as if tab is intended.  Will fix.

File: sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+		((ExperimentConfigImpl)experimentConfig).setPeriod((float) currentGoDirectDevice.getMeasurementPeriod());
+
+		return experimentConfig;
+	}
+
+	public ExperimentConfig getCurrentConfig() {
+		if(currentGoDirectDevice != null){
+			currentGoDirectDevice.close();
+			currentGoDirectDevice = null;
+		}
```
[the remark this answers is not in the document]
Comment (mklewandowski) [reply sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java]: This pattern is taken from GoIO.  Close and reopen in the getCurrentConfig.  I agree, it seems odd.  Is the assumption that the an incorrect device might be open?

File: sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+public class D2PIOSensorDevice extends AbstractSensorDevice implements
+		VernierSensorDevice
+{
+	D2PIOLibrary d2pio;
+	String errorMessage;
+	D2PIOSensor currentGoDirectDevice;
+  private static final int MAX_D2PIO_CHANNELS = 32;
+
+	public D2PIOSensorDevice() {
+		deviceLabel = "GODIR";
```
[the remark this answers is not in the document]
Comment (mklewandowski) [reply sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java]: The standard was actually 2-3 characters so I might actually be over an expected limit here.  Internally, other than logging, I don't see this used much.  Not sure who eventually might use it.  "GDX" is commonly seen in the code provided by Vernier.  I vote to move to that.

File: sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java (hunk 1 above)
Comment (mklewandowski) [reply sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java]: Device label appears to be a shortened version of the device name (2-3 char).  I'm voting "GDX" for device label and "Go Direct" with a space for device name.  The space is how Vernier does it on their promotional materials:
https://www.vernier.com/products/sensors/go-direct-sensors/

**Pull request merged**